### ISSUE TITLE
how train

### ISSUE THREAD

**Issue by mamadoudelacitenoir** [ultralytics/ultralytics#4947]
### Search before asking

- [ ] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

How i can trained a onnx model with a python script, i  have good images and wrong images and i want create a onnx model with that for object detections (yolov8) and all that with a python script, ty.

i want train a model only with images of object , and i have bad images to don't detect too

**Comment by glenn-jocher** [comment]
@mamadoudelacitenoir thank you for reaching out and showing interest in training a model with YOLOv8.

Before you start the training process, you will need to prepare your dataset in a certain format. A typical object detection dataset would include images labeled with objects' bounding boxes. The dataset should include both positive examples (images containing the objects of interest) and negative examples (images not containing any objects of interest). Both good images and wrong images, as you refer to them, are important for robust model training.

Here is a general step-by-step guide to training YOLOv8 with a custom dataset:

1. Dataset Preparation: Arrange your images into folders and label them. This usually involves providing a bounding box around the objects in the image and assigning a class label to each bounding box.

2. Dataset Configuration: Create a dataset configuration file. This is a YAML file where you will specify the paths to the training and validation datasets and number of classes.

3. Training Configuration: Create a model configuration file. This is another YAML file where you will specify the architecture settings like the number and types of layers, loss parameters, etc.

4. Model Training: Use the 'train' command along with the path to the model configuration file, dataset configuration file, and weights.

5. Validation and Error Analysis: After the training is done, use the 'val' command to validate the trained model on your validation dataset. This should help you identify any potential issues with your model.

Given that you want to export your model in the ONNX format, you can use the 'export' mode after you have trained your model. This will convert the PyTorch format model to ONNX format, which can then be used for deployment.

Please refer to the YOLOv8 documentation for specifics on how to create these configuration files and the exact commands for training and validation.

Note: You can carry out all these steps in a Python script by calling the relevant YOLOv8 functions.

Good luck with your model training! Let us know if you have more questions or face any issues, we're here to help.

**Comment by mamadoudelacitenoir** [comment]
hi, you said i need labels them, but i alrd have a lot imgs of the object, so i don't need label it

**Comment by glenn-jocher** [comment]
@mamadoudelacitenoir hello, thank you for your query.

Training a model for object detection, like YOLOv8, involves teaching the model to recognize certain objects within images and to identify their locations. The way we teach the model this is through labeled data. Labeled images come with additional information, typically in the form of bounding boxes and class labels that indicate where the objects are and what categories they belong to.

If you don't have labeled data, the model won't know what to learn from your images. It won't be able to differentiate between the object of interest and the background or other objects. Essentially, the labels serve as the teacher to our model student.

If you only have images without labels and you want to use them for training anyway, you may want to look into unsupervised learning methods, or semi-supervised learning if you have a small portion of labeled images. However, note that YOLOv8 is a supervised learning algorithm and it requires labeled data for training.

Hopefully, this clarifies the importance and role of labels in the training process for YOLOv8 model. If you have additional questions or if there’s anything else you’d like to know, feel free to ask.

**Comment by mamadoudelacitenoir** [comment]
then there is others yolov8 can work for what i want do ?

**Comment by glenn-jocher** [comment]
@mamadoudelacitenoir if you're trying to train a model using images without labels, you might want to consider an unsupervised learning approach. It should be noted, however, that YOLOv8 is a supervised learning algorithm and it requires labeled data for training. 

Unsupervised learning methods, such as clustering, can be used when you don't have labeled data. These methods discover hidden patterns or data groupings without the need for human intervention. 

Another approach could be training an autoencoder. Autoencoders are a type of artificial neural network used to learn efficient data codings in an unsupervised manner. They can be trained to copy their input to their output, with the aim of learning a representation (encoding) for a set of data.

These methods may provide a path forward given that you only have access to unlabeled data. If you do manage to get some labelled data later on, you'll likely want to switch back to a supervised learning algorithm like YOLOv8, as these generally provide better performance for tasks such as object detection. 

Hopefully this provides some clarity on the different types of learning algorithms available to you depending on the type of data you have to work with. If you have further questions, please do not hesitate to ask.

**Comment by mamadoudelacitenoir** [comment]
i have one idea, i have only good images , so this mean all the image is right, labeling = get  x1 ,y1 & x2 , y2 
and me i have all right images so labeling = size of images no ? so i can do a auto script for labeling & train right ?

**Comment by glenn-jocher** [comment]
@mamadoudelacitenoir hello, thanks for reaching out with your query.

If I understand correctly, you're asking if you can label an entire image as a single bounding box for the object of interest. Technically, you can create a script to automatically label all images using the image size as the bounding box coordinates. It will label the full image as containing an instance of your object, if you are certain that every image only contains the object of interest and nothing else.

This automatically labeled data can then potentially be used to train a model like YOLOv8. However, it's worth noting that object detection models like YOLOv8 are traditionally utilized for detecting instances of various objects within the same image and outputting their respective bounding boxes.

If your images each contain a single object that takes up the entire image, and you only have a single class, you might not be leveraging the full capabilities of object detection models like YOLOv8. You may want to explore image classification models instead, which can predict whether a given object is present in the image.

It's important to pick the right tool for your project's needs, so do consider thoroughly what kind of tasks you aim to achieve with your model. 

I hope this answers your question. If you have more queries, feel free to ask.

**Comment by mamadoudelacitenoir** [comment]
ty for all the help, im making a script for auto train.

**Comment by mamadoudelacitenoir** [comment]
how i can train a model with a amd gpu

**Comment by glenn-jocher** [comment]
@mamadoudelacitenoir hello,
Thanks for your question. The YOLOv8 model, as with many deep learning models, is typically trained using GPUs for efficient computation - specifically NVIDIA GPUs because they support CUDA, an API that allows for performing general purpose computations on GPUs.

Unfortunately, AMD GPUs do not natively support CUDA because it's a proprietary framework by NVIDIA. However, there's an open-source platform called ROCm (Radeon Open Compute) developed by AMD for running computations on AMD GPUs. Some deep learning frameworks have started supporting ROCm to enable training models on AMD GPUs. 

Before proceeding, please check if the deep learning framework you're using (like TensorFlow, PyTorch, etc.) supports ROCm. If it does, you'll have to install ROCm and configure your deep learning environment to use it. 

Please note that while ROCm is a step in the right direction, working with AMD GPUs for deep learning tasks might be limiting due to the widespread use and support of CUDA in many popular libraries and frameworks.

Remember, optimal training of deep learning models often necessitates specific hardware, so using a CPU or GPU that is not fully compatible with the system requirements may lead to underperformance, errors, or inability to train the model.

Hope this helps! Let us know if you have more questions.

**Comment by swarupbc** [comment]
Hi @glenn-jocher, I'm trying to train a model using ultralytics and pytorch on my AMD Radeon Rx7600, Ubuntu distro, and rocm installed.

here is the code from my ipynb file - 

!rocm-smi

%pip install ultralytics

from IPython import display
display.clear_output()

import ultralytics

ultralytics.checks()

from ultralytics import YOLO

from IPython.display import display,Image

%cd updated_mango_disease_dataset_01032024

!yolo task=detect mode=train model=yolov8x.pt data=data_training.yaml epochs=300 plots=True batch=16 patience=0

it's suppose to use Nvidia CUDA, but as I have an AMD GPU with rocm installed, how can i make it to use rcom instead of CUDA?

**Comment by glenn-jocher** [comment]
Hey there! 😊 It looks like you're making great strides with your setup. To focus on ROCm for AMD GPUs instead of CUDA, you might need to ensure a couple of things are in place since the primary support of Ultralytics YOLO models and PyTorch is geared towards CUDA.

Firstly, verify that your PyTorch installation is ROCm-compatible. PyTorch offers support for ROCm but usually requires a specific installation command. You might need to reinstall or create a new environment with the ROCm-compatible PyTorch version. Here's an example to install PyTorch for ROCm:

```bash
pip install torch torchvision --extra-index-url https://download.pytorch.org/whl/rocm4.2
```
Please replace `rocm4.2` with the version compatible with your ROCm installation. You can check the PyTorch website or PyTorch GitHub page for the exact command for your ROCm version.

Secondly, once you have the ROCm version of PyTorch, most operations should automatically use ROCm instead of CUDA. PyTorch abstracts most of the hardware-specific calls, so it'll automatically utilize your AMD GPU if it's supported and correctly set up with ROCm.

Unfortunately, not all CUDA functionalities may be fully supported or directly translate to ROCm, so you might encounter certain limitations or the need for workarounds based on what specific features your training leverages.

If you run into specific errors or functionalities not behaving as expected, I'd recommend checking the issues section of the PyTorch GitHub repo, or even the Ultralytics repo, as others may have encountered similar situations.

Hope this nudges you in the right direction! If you hit any roadblocks, feel free to ask. Happy coding! 🚀
